src/openrct2/localisation/FormatCodes.cpp

```diff
@@ -9,116 +9,175 @@
+    return result != std::end(FormatTokenMap) ? result->second : FormatToken::Unknown;
+}
+
+std::string_view FormatTokenToString(FormatToken token)
+{
+    for (const auto& t : FormatTokenMap)
     {
-        if (_strcmpi(token, format_code_tokens[i].token) == 0)
-            return format_code_tokens[i].code;
+        if (t.second == token)
+        {
+            return t.first;
+        }
     }
-    return 0;
+    return {};
 }
 
-const char* format_get_token(uint32_t code)
+bool FormatTokenTakesArgument(FormatToken token)
 {
-    for (uint32_t i = 0; i < std::size(format_code_tokens); i++)
+    switch (token)
     {
-        if (code == format_code_tokens[i].code)
-            return format_code_tokens[i].token;
+        case FormatToken::Comma32:
+        case FormatToken::Int32:
+        case FormatToken::Comma1dp16:
+        case FormatToken::Comma2dp32:
+        case FormatToken::Comma16:
+        case FormatToken::UInt16:
+        case FormatToken::Currency2dp:
+        case FormatToken::Currency:
+        case FormatToken::StringId:
+        case FormatToken::String:
+        case FormatToken::MonthYear:
+        case FormatToken::Month:
+        case FormatToken::Velocity:
+        case FormatToken::DurationShort:
+        case FormatToken::DurationLong:
+        case FormatToken::Length:
+        case FormatToken::Sprite:
+            return true;
+        default:
+            return false;
     }
-    return nullptr;
 }
 
-bool utf8_should_use_sprite_for_codepoint(char32_t codepoint)
+bool FormatTokenIsColour(FormatToken token)
 {
-    switch (codepoint)
+    switch (token)
     {
-        case UnicodeChar::up:
-        case UnicodeChar::down:
-        case UnicodeChar::leftguillemet:
-        case UnicodeChar::tick:
-        case UnicodeChar::cross:
-        case UnicodeChar::right:
-        case UnicodeChar::rightguillemet:
-        case UnicodeChar::small_up:
-        case UnicodeChar::small_down:
-        case UnicodeChar::left:
-        case UnicodeChar::quote_open:
-        case UnicodeChar::quote_close:
-        case UnicodeChar::german_quote_open:
-        case UnicodeChar::plus:
-        case UnicodeChar::minus:
+        case FormatToken::ColourBlack:
+        case FormatToken::ColourGrey:
+        case FormatToken::ColourWhite:
+        case FormatToken::ColourRed:
+        case FormatToken::ColourGreen:
+        case FormatToken::ColourYellow:
+        case FormatToken::ColourTopaz:
+        case FormatToken::ColourCeladon:
+        case FormatToken::ColourBabyBlue:
+        case FormatToken::ColourPaleLavender:
+        case FormatToken::ColourPaleGold:
+        case FormatToken::ColourLightPink:
+        case FormatToken::ColourPearlAqua:
+        case FormatToken::ColourPaleSilver:
             return true;
         default:
             return false;
     }
 }
 
-#pragma endregion
+size_t FormatTokenGetTextColourIndex(FormatToken token)
+{
+    switch (token)
+    {
+        case FormatToken::ColourBlack:
```

Comment:
Is this (roughly) as efficient as just doing a bounds check and subtracting?

Reply:
Format tokens are now in no particular order, you can change it freely and things still work. But it requires switch like this. There was already a debate between switch and static array in a different PR, switch ensure no extra changes are required when the values of FormatToken change.